crates/air-lib/air-parser/src/parser/validator.rs

```diff
@@ -30,6 +30,208 @@ use multimap::MultiMap;
+    fn disable(&mut self) {
+        self.stack.clear();
+        self.is_enabled = false;
+    }
+
+    fn malformed_spans_iter(&self) -> std::slice::Iter<Span> {
+        self.malformed_spans.iter()
+    }
+
+    fn met_instruction_kind(&mut self, instr_kind: CheckInstructionKind<'name>, span: Span) {
+        use CheckInstructionKind::*;
+
+        if !self.is_enabled {
+            return;
+        }
+        match instr_kind {
+            CheckInstructionKind::Replacing => {
+                self.process_replacing(span);
+            }
+            CheckInstructionKind::Xoring => {
+                self.process_xoring(span);
+            }
+            CheckInstructionKind::Merging => {
+                self.process_merging(span);
+            }
+            CheckInstructionKind::ReplacingWithCheck(iterator_name) => {
+                self.replacing_with_check_common(iterator_name, span, &instr_kind);
+            }
+            PopStack1ReplacingWithCheck(iterator_name) => {
+                self.stack.pop();
+                self.replacing_with_check_common(iterator_name, span, &instr_kind);
+            }
+            PivotalNext(_) | Simple => self.stack.push((instr_kind, span)),
+            PopStack1 => {
+                self.stack.pop();
+                self.stack.push((instr_kind, span))
+            }
+            PopStack2 => {
+                self.stack.pop();
+                self.stack.pop();
+                self.stack.push((instr_kind, span))
+            }
+        }
+    }
+
+    fn process_replacing(&mut self, span: Span) {
+        use CheckInstructionKind::*;
+
+        let child = self.stack.pop();
+        match child {
+            Some((pattern_kind @ PivotalNext(_), ..)) => {
+                self.stack.push((pattern_kind, span));
+            }
+            Some(_) => {
+                self.stack.push((Replacing, span));
+            }
+            None => self.disable(),
+        }
+    }
+
+    fn process_xoring(&mut self, span: Span) {
+        use CheckInstructionKind::*;
+
+        let right_branch = self.stack.pop();
+        let left_branch = self.stack.pop();
+        let left_right = left_branch.zip(right_branch);
+
+        match left_right {
+            Some(((PivotalNext(_), ..), (PivotalNext(_), ..))) => {
+                // `xor` has `next` in both branches. It is impossible
+                // to tell which branch should poped up.
+                self.disable();
+            }
+            Some(((pattern_kind @ PivotalNext(_), ..), ..))
+            | Some((_, (pattern_kind @ PivotalNext(_), _))) => {
+                // potential failure but need to check when fold pops up.
+                self.stack.push((pattern_kind, span));
+            }
+            Some(_) => {
+                self.stack.push((Xoring, span));
+            }
+            _ => {
+                // disable machine if Xoring invariant, namely there must be 2 kinds on a stack, is broken.
```

Comment:
Can the invariant broken, but there are still 2 kinds on a stack, one is from some unrelated statement?

Does this invariant relates to correctness of parsing (i.e. it can break only if parsing/parser is incorrect)?

Reply:
No.
Yes. Changed semantics might also result in a failure.